File: TestApp/src/main/java/br/com/insanitech/testapp/TestAppActivity.java
package br.com.insanitech.testapp;

import android.app.Activity;
import android.os.Bundle;
import android.view.MotionEvent;
import android.view.View;

import java.util.Arrays;

import br.com.insanitech.spritekit.SKNode;
import br.com.insanitech.spritekit.SKScene;
import br.com.insanitech.spritekit.SKSpriteNode;
import br.com.insanitech.spritekit.SKTexture;
import br.com.insanitech.spritekit.SKView;
import br.com.insanitech.spritekit.actions.SKAction;
import br.com.insanitech.spritekit.graphics.SKColor;
import br.com.insanitech.spritekit.graphics.SKRect;
import br.com.insanitech.spritekit.graphics.SKSize;
import kotlin.Unit;
import kotlin.jvm.functions.Function0;

/**
 * Created by anderson on 24/06/15.
 */
public class TestAppActivity extends Activity implements View.OnTouchListener {
    private SKView view;
    private SKTexture texture;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
    }

    @Override
    protected void onPause() {
        super.onPause();

        if (view != null) {
            view.onPause();
        }
    }

    @Override
    protected void onResume() {
        super.onResume();

        if (view != null) {
            view.onResume();
        }
    }

    @Override
    public void onWindowFocusChanged(boolean hasFocus) {
        super.onWindowFocusChanged(hasFocus);

        // call this method here so the views are ready,
        // otherwise the width and height properties of all components will return 0;
        initializeScene();
    }

    private void initializeScene() {
        if (view == null) {
            view = (SKView) findViewById(R.id.skview);
            SKScene scene = new SKScene(new SKSize(320, 480));
            scene.setBackgroundColor(SKColor.Companion.darkGray());

            texture = new SKTexture(this, R.drawable.card_deck);

            SKSpriteNode swordA = SKSpriteNode.Companion.spriteNode(SKColor.Companion.white(), new SKSize(100, 200));
            swordA.setTexture(new SKTexture(new SKRect(0.0f, 0.0f, 1.0f / 14.0f, 1.0f / 4.0f), texture));
            swordA.setColorBlendFactor(0.5f);

//            SKSpriteNode heartA = SKSpriteNode.spriteNode(SKColor.white(), new SKSize(320, 480));
//            heartA.setTexture(new SKTexture(new SKRect(0.0f, 2.0f / 4.0f, 1.0f / 14.0f, 1.0f / 4.0f), texture));
//            heartA.setColorBlendFactor(0.5f);

            SKNode nodeParent = SKNode.Companion.node();
            nodeParent.getPosition().setX((scene.getSize().getWidth() / 2.0f));
            nodeParent.getPosition().setY((scene.getSize().getHeight() / 2.0f));

            nodeParent.addChild(swordA);
//            nodeParent.addChild(heartA);

//            heartA.zPosition = -1;

            scene.addChild(nodeParent);

            view.presentScene(scene);

            view.setOnTouchListener(this);

//            animateMovingCards(swordA);

            swordA.setZRotation(0);
            swordA.run(SKAction.Companion.sequence(Arrays.asList(SKAction.Companion.waitFor(1000), SKAction.Companion.rotateToAngle((float) (Math.PI / 2), 500))));
            swordA.run(SKAction.Companion.sequence(Arrays.asList(SKAction.Companion.waitFor(2000), SKAction.Companion.rotateToAngle((float) Math.PI, 500))));
            swordA.run(SKAction.Companion.sequence(Arrays.asList(SKAction.Companion.waitFor(3000), SKAction.Companion.rotateToAngle((float) (3 * Math.PI / 2), 500))));
            swordA.run(SKAction.Companion.sequence(Arrays.asList(SKAction.Companion.waitFor(4000), SKAction.Companion.rotateToAngle((float) (Math.PI * 2), 500))));
        }
    }

    private void rotate(final SKNode nodeParent) {
        nodeParent.run(SKAction.Companion.sequence(Arrays.asList(
                SKAction.Companion.rotateByAngle(1, 50),
                SKAction.Companion.run(new Function0<Unit>() {
                    @Override
                    public Unit invoke() {
                        rotate(nodeParent);
                        return null;
                    }
                }))));
    }

    @Override
    public boolean onTouch(View v, MotionEvent event) {
        switch (event.getAction()) {
            case MotionEvent.ACTION_MOVE:

                return true;

            default:
                break;
        }
        return false;
    }

    private void animateMovingCards(final SKSpriteNode sprite) {
        float spriteCoordX = sprite.getTexture().textureRect().getX();
        float spriteCoordY = sprite.getTexture().textureRect().getY();

        spriteCoordX += 1.0f / 14.0f;

        if (spriteCoordX >= 1.0f) {
            spriteCoordX = 0.0f;
            spriteCoordY += 1.0f / 4.0f;
        }

        if (spriteCoordY >= 1.0f) {
            spriteCoordX = 0.0f;
            spriteCoordY = 0.0f;
        }

        sprite.run(SKAction.Companion.sequence(Arrays.asList(SKAction.Companion.waitFor(1000),
                SKAction.Companion.setTexture(new SKTexture(new SKRect(spriteCoordX, spriteCoordY, 1.0f / 14.0f, 1.0f / 4.0f), texture)),
                SKAction.Companion.run(new Function0<Unit>() {
                    @Override
                    public Unit invoke() {
                        animateMovingCards(sprite);
                        return null;
                    }
                }))));
    }
}

File: TestApp/src/main/java/br/com/insanitech/testapp/TestAppActivity.kt
package br.com.insanitech.testapp

import android.app.Activity
import android.os.Bundle
import android.view.MotionEvent
import android.view.View
import br.com.insanitech.spritekit.SKNode
import br.com.insanitech.spritekit.SKScene
import br.com.insanitech.spritekit.SKSceneScaleMode.AspectFill
import br.com.insanitech.spritekit.SKSpriteNode
import br.com.insanitech.spritekit.SKView
import br.com.insanitech.spritekit.graphics.SKColor
import br.com.insanitech.spritekit.graphics.SKPoint
import br.com.insanitech.spritekit.graphics.SKSize
import kotlinx.android.synthetic.main.activity_main.*
import kotlin.system.exitProcess

/**
 * Created by anderson on 24/06/15.
 */
class TestAppActivity : Activity(), View.OnTouchListener {
    val view: SKView
        get() = this.skView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
    }

    override fun onPause() {
        super.onPause()

        this.view.onPause()
    }

    override fun onResume() {
        super.onResume()

        this.view.onResume()
    }

    override fun onWindowFocusChanged(hasFocus: Boolean) {
        super.onWindowFocusChanged(hasFocus)

        // call this method here so the views are ready,
        // otherwise the width and height properties of all components will return 0;
        this.skView.queueEvent {
            this.initializeScene()
        }
    }

    private fun initializeScene() {
        val scene = SKScene(SKSize(320f, 568f))
        scene.name = "Scene"
        scene.scaleMode = AspectFill
        scene.backgroundColor = SKColor.darkGray()

        this.view.presentScene(scene)
        this.view.setOnTouchListener(this)

        val parentNode = SKNode.node()
        parentNode.name = "Parent"
        parentNode.position = SKPoint(scene.size.width / 2.0f, scene.size.height / 2.0f)
        scene.addChild(parentNode)

        var spriteNode = SKSpriteNode.spriteNode(SKColor.red(), SKSize(50f, 50f))
        spriteNode.position.x -= 20
        spriteNode.name = "Red SpriteNode"
        parentNode.addChild(spriteNode)

        System.out.println("accum frame red: ${spriteNode.calculateAccumulatedFrame()}")

        spriteNode = SKSpriteNode.Companion.spriteNode(SKColor.blue(), SKSize(50f, 50f))
        spriteNode.position.x += 20
        spriteNode.name = "Blue SpriteNode"
        spriteNode.zPosition = -1f
        parentNode.addChild(spriteNode)

        spriteNode = SKSpriteNode.Companion.spriteNode(SKColor.white(), scene.size)
        spriteNode.name = "Background Node"
        spriteNode.zPosition = -100f
        spriteNode.anchorPoint = SKPoint(0f, 0f)
        scene.addChild(spriteNode)

        System.out.println("accum frame blue: ${spriteNode.calculateAccumulatedFrame()}")

        System.out.println("accum frame parent: ${parentNode.calculateAccumulatedFrame()}")
    }

    override fun onTouch(v: View, event: MotionEvent): Boolean {
        System.out.println(event.toString())
        when (event.actionMasked) {
            MotionEvent.ACTION_UP -> {
                this.view.queueEvent {
                    val touchPosition = SKPoint(event.x, event.y)
                    val scenePosition = this.view.convertTo(touchPosition, this.view.scene!!)

                    val parent = this.view.scene!!.children[0]
                    var convertedPoint = parent.convertFrom(scenePosition, this.view.scene!!)

                    val nodeFromParent = parent.atPoint(convertedPoint)

                    convertedPoint = this.view.scene!!.convertFrom(convertedPoint, parent)
                    val nodeFromScene = this.view.scene!!.atPoint(convertedPoint)

                    if (nodeFromParent != nodeFromScene) {
                        exitProcess(1)
                    }
                    System.out.println(nodeFromParent.name)
                    System.out.println(nodeFromScene.name)
                }
            }
            else -> { }
        }
        return true
    }
}
